### PR TITLE
feat: add p/demo/uassert & p/demo/urequire

### PR DESCRIPTION
**Pull request by moul**
Addresses #923 
Addresses #777

### CODE REVIEW

File: examples/gno.land/p/demo/ownable/ownable_test.gno [hunk 1]
```diff
@@ -67,9 +60,7 @@ func TestCallerIsOwner(t *testing.T) {
 	std.TestSetOrigCaller(unauthorizedCaller) // TODO(bug): should not be needed
 
 	err := o.CallerIsOwner()
-	if err == nil {
-		t.Fatalf("Expected %s to not be owner", unauthorizedCaller)
-	}
+	uassert.Error(t, err) // XXX: IsError(..., unauthorizedCaller)
```
Comment (ajnavarro) [inline examples/gno.land/p/demo/ownable/ownable_test.gno:63]: Why implementing IsError is not possible?

File: examples/gno.land/p/demo/uassert/uassert.gno [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+func False(t TestingT, value bool, msgs ...string) bool {
+	t.Helper()
+	if value {
+		return fail(t, msgs, "should be false")
+	}
+	return true
+}
+
+// ErrorIs asserts the given error matches the target error
+func ErrorIs(t TestingT, err, target error, msgs ...string) bool {
```
Comment (ajnavarro) [inline examples/gno.land/p/demo/uassert/uassert.gno:64]:
```suggestion
func IsError(t TestingT, err, target error, msgs ...string) bool {
```

File: examples/gno.land/p/demo/urequire/urequire.gno [hunk 3]
```diff
@@ -0,0 +1,87 @@
+// urequire is a sister package for uassert.
+// XXX: codegen the package.
+package urequire
+
+import "gno.land/p/demo/uassert"
+
+// type TestingT = uassert.TestingT // XXX: bug, should work
```
Comment (ajnavarro) [inline examples/gno.land/p/demo/urequire/urequire.gno:7]: What is exactly the bug here? do we have an issue that we can add here to track it somehow?

**Review by ajnavarro — APPROVED** [review]
LGTM after addressing some minor comments

File: examples/gno.land/p/demo/ownable/ownable_test.gno (hunk 1 above)
Comment (zivkovicmilos) [reply examples/gno.land/p/demo/ownable/ownable_test.gno:63]: The base go implementation of this uses reflection 🙁

File: examples/gno.land/p/demo/uassert/uassert.gno (hunk 2 above)
Comment (zivkovicmilos) [reply examples/gno.land/p/demo/uassert/uassert.gno:64]: `ErrorIs` is the wording in the assert package, which this is trying to mimic
https://pkg.go.dev/github.com/stretchr/testify/assert#ErrorIs

File: examples/gno.land/p/demo/uassert/uassert_test.gno [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+		{"Hello World", "Hello World", true, ""},
+		{123, 123, true, ""},
+		{123.5, 123.5, true, ""},
+		{nil, nil, true, ""},
+		{int32(123), int32(123), true, ""},
+		{uint64(123), uint64(123), true, ""},
+		{std.Address("g12345"), std.Address("g12345"), true, ""},
+		//{[]byte("Hello World"), []byte("Hello World"), true, ""},
+		// {myType("1"), myType("1"), true, ""},
+		//{&struct{}{}, &struct{}{}, true, "pointer equality is based on equality of underlying value"},
```
Comment (zivkovicmilos) [inline examples/gno.land/p/demo/uassert/uassert_test.gno:138]: Leftover?

File: examples/gno.land/p/demo/uassert/uassert.gno [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+
+	/*
+		type stringer interface{ String() string }
+		if ev, ok := expected.(stringer); ok {
+			if av, ok := actual.(stringer); ok {
+				equal = ev.String() == av.String()
+				ok_ = true
+			}
+		}
+	*/
```
Comment (zivkovicmilos) [inline examples/gno.land/p/demo/uassert/uassert.gno:229]: Leftover?

**Review by zivkovicmilos — APPROVED** [review]
I need this in my life. One step closer to making gno usable on the day-to-day

File: examples/gno.land/p/demo/ownable/ownable_test.gno (hunk 1 above)
Comment (moul) [reply examples/gno.land/p/demo/ownable/ownable_test.gno:63]: we can do ErrorContains(t, err, myerr.Error()) which is using strings.Contains.

Comment (moul) [reply examples/gno.land/p/demo/ownable/ownable_test.gno:63]: The previous test was simply checking that the error is not equal to nil, which is exactly what `uassert.Error` is doing. My comment is more of a suggestion to improve the unit tests in the future (comparing error type, not just non-nil). However, I believe we are equally good or better at testing code as before with this PR.

File: examples/gno.land/p/demo/uassert/uassert_test.gno (hunk 4 above)
Comment (moul) [reply examples/gno.land/p/demo/uassert/uassert_test.gno]: c621d6ef..c2622b99

File: examples/gno.land/p/demo/uassert/uassert.gno (hunk 5 above)
Comment (moul) [reply examples/gno.land/p/demo/uassert/uassert.gno]: a1cd5502

File: examples/gno.land/p/demo/urequire/urequire.gno (hunk 3 above)
Comment (zivkovicmilos) [reply examples/gno.land/p/demo/urequire/urequire.gno:7]: I remember seeing this _exact_ issue a while back in local dev, but forgot to open up an issue for it 🤦‍♂️
